### PR TITLE
CSL4LibreOffice - C [GSoC '24]

### PR DESCRIPTION
**Pull request by subhramit**
# New features for CSL-OO/LO integration
**[PR - C of GSoC '24 CSL4LibreOffice Project]**
Follow-up to https://github.com/JabRef/jabref/pull/11521

## Background:
- **Problem 1:** The behavior of in-text citations was independent of bibliographic citations, even if they corresponded to the same entry. For example, they could have different numbering (for numeric styles).
- **Problem 2:** As mentioned by @Siedlerchr in https://github.com/JabRef/jabref/pull/11521#issuecomment-2246323544, a major inconvenience while citing entries independently was that the numbering index (for numeric citation styles) would get reset to 1 every time.
- **Problem 3:** We first tried ways to persist consistent numbering within an OO session, the numbering would get reset again on restarting JabRef or reconnecting/restarting the LibreOffice document instance.
- **Problem 4:** There was no availability of a "make bibliography" feature that would automatically read the in-text citations in the document and generate a bibliography consisting of corresponding bibliographic citations.

This PR addresses these problems by introducing two new features to JabRef's OO/LO integration - (A) **Reference Marks** and (B) **Auto-generation of bibliography**.
## Changes:
- We can now seamlessly insert citation entries (both independently or in a grouped fashion) without having to worry about maintaining the numbering across citations of any type. The numbers are maintained consistently using Reference Marks attached to citation entries.
- We can now auto-insert bibliography into the document based on our in-text citations by pressing the loop arrow ("Make bibliography") button in the OpenOffice panel.

https://github.com/user-attachments/assets/c5a1c338-b42e-421a-a86f-69a06f418eab
- The citation numbers are now maintained consistently across JabRef and OO/LO sessions by reading existing Reference Marks in the document.

https://github.com/user-attachments/assets/0b59b38a-eaca-4c33-8b55-429fce7d44c7

- The implementation of the internal format of Reference Marks is very similar to that of Zotero, which consists of "`JABREF_{citationKey} CID_{citationNumebr} {UID}`".

Special thanks to @antalk2, who had worked on similar problem statements for JStyles for JabRef's OO/LO integration, for his help in guiding and making this PR possible.
### Mandatory checks

- [ ] ~Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)~
- [ ] ~Tests created for changes (if applicable)~
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by subhramit** [comment]
TODOs (before merge):
1. Possible refinement of implementation over the remaining weeks, based on reviews and testing.
2. Documentation of code and clean up old comments.

File: src/main/java/org/jabref/gui/openoffice/OOBibBase.java
Comment (calixtus) [inline src/main/java/org/jabref/gui/openoffice/OOBibBase.java:960]: It's a bad Idea.

The injector currently cares about singletons across the whole application. So injecting a BibDatabaseContext would assume, that there is only one BibDatabaseContext in the whole application. But there are multiple. one for each opened BibDatabase.

File: src/main/java/org/jabref/gui/openoffice/OpenOfficePanel.java
Comment (calixtus) [inline src/main/java/org/jabref/gui/openoffice/OpenOfficePanel.java:237]: this needs to be dependend on the observable. The currentStyle can switch between JStyle and CSL without the OpenOfficePanel reloading.

File: src/main/java/org/jabref/gui/openoffice/OOBibBase.java
Comment (calixtus) [inline src/main/java/org/jabref/gui/openoffice/OOBibBase.java:952]: this should not use hardcoded linebreaks or fontweights or font sizes, but paragraph styles. See OOBibStyle#getFormattedBibliographyTitle

File: src/main/java/org/jabref/logic/openoffice/oocsltext/CSLReferenceMark.java
Comment (koppor) [inline src/main/java/org/jabref/logic/openoffice/oocsltext/CSLReferenceMark.java:59]: This is unused. Strange.

File: src/main/java/org/jabref/logic/openoffice/oocsltext/CSLReferenceMark.java
Comment (koppor) [inline src/main/java/org/jabref/logic/openoffice/oocsltext/CSLReferenceMark.java:63]: This is unused. Strange.

File: src/main/java/org/jabref/logic/openoffice/oocsltext/CSLReferenceMark.java
Comment (subhramit) [reply src/main/java/org/jabref/logic/openoffice/oocsltext/CSLReferenceMark.java:59]: I made the getters first, later on they weren't needed but still kept them for future extensibility (looked like we might).

Will remove them right now if you want.

File: src/main/java/org/jabref/logic/openoffice/oocsltext/CSLCitationOOAdapter.java
Comment (koppor) [inline src/main/java/org/jabref/logic/openoffice/oocsltext/CSLCitationOOAdapter.java:24]: Is the prefix `CSL_` really needed? I think, only `JABREF_` is written as prefix.

**Comment by subhramit** [comment]
> TODOs (before merge):
> 1. Possible refinement of implementation over the remaining weeks, based on reviews and testing.
> 2. Documentation of code (javadoc) and clean up old comments and logs.
> 3. Fix submodules

TODOs (independent/follow-up PR)
1. Add tests for PR - A, B (`transformHTML`, OO HTML generation).
2. Add tests for PR - C (reference mark generation, parsing and reading).

File: src/main/java/org/jabref/gui/openoffice/OOBibBase.java
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/openoffice/OOBibBase.java:124]: duplicated code to  initializeCitationAdapter?

File: src/main/java/org/jabref/gui/openoffice/OOBibBase.java
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/openoffice/OOBibBase.java:644]: Never catch all exceptions

File: src/main/java/org/jabref/gui/openoffice/OOBibBase.java
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/openoffice/OOBibBase.java:838]: Not sure if this works, I think there was something about using vars for l10n https://devdocs.jabref.org/code-howtos/localization.html#general-hints

File: src/main/java/org/jabref/logic/citationstyle/CitationStyle.java
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/citationstyle/CitationStyle.java:244]: Document Builder is costly to be initalied, make it a static const     private static final DocumentBuilderFactory DOCUMENT_BUILDER_FACTORY = DocumentBuilderFactory.newInstance();

Comment (Siedlerchr) [reply src/main/java/org/jabref/logic/citationstyle/CitationStyle.java:244]: Ideally, this should be in the citeproc java

File: src/main/java/org/jabref/logic/citationstyle/CitationStyle.java
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/citationstyle/CitationStyle.java:256]: use slf4.logger

**Comment by subhramit** [comment]
Fixes https://github.com/subhramit/jabref/issues/16

**Comment by subhramit** [comment]
@calixtus @koppor @Siedlerchr @ThiloteE Decision to leave the bibliography title format upto the user:
In "Make bibliography" of Jstyles, the generated bibliography section has a heading, whose text and formatting specifications (including paragraph styles) are variable and taken from the jstyle that specifies it. This was, hence, indirectly left to the user who would create the jstyle.

Since CSL citations do not specify any such formatting (I checked), if we add a default pre-formatted bibliography title, it can be inconvenient/irritating for a wide range of users as they will have to remove it and put a title of their own, since we would make the following assumptions:
1. Title text - whether they want to call it "References" or "Bibliography" or something else
2. Where to put the title - Align left? Or center? Or in rare cases, even right?
3. Font weight - Bold/plain
4. Font size - So many choices. Would our default one be coherent/suitable as per the text size in the rest of the document?
5. 5. Paragraph style - do they want a specific paragraph style (heading)?
6. Italicized? Underlined? Other formatting?

Since just adding a reference heading is not much of a task, I am in favor of avoiding putting a title at all, leave it to the user, and just generate the bibliography section.
image

**Comment by subhramit** [comment]
Extra newline issue for numeric styles fixed.
image

**Comment by subhramit** [comment]
This partially fixes https://github.com/subhramit/jabref/issues/13 (2/3). The third requirement is out of scope and has to be filed upstream.

File: src/main/java/org/jabref/gui/openoffice/OpenOfficePanel.java
Comment (subhramit) [reply src/main/java/org/jabref/gui/openoffice/OpenOfficePanel.java]: No longer required as refresh feature implemented for CSL as well. Will have a unified tooltip now.

**Comment by subhramit** [comment]
Fixes https://github.com/subhramit/jabref/issues/18
Styles having underline are now supported.

image

**Comment by subhramit** [comment]
TODO: Rewrite pr description as per updates.

**Comment by koppor** [comment]
> @calixtus @koppor @Siedlerchr @ThiloteE Decision to leave the bibliography title format upto the user:

TL;DR; You are asking how to format the heading.

User exit: Write the bibliography without header. The user has to insert it.

Alternatively: Use LibreOffice style "Heading 2", which should always be available. This is what formatting styles are made for.

**Comment by calixtus** [comment]
> @calixtus @koppor @Siedlerchr @ThiloteE Decision to leave the bibliography title format upto the user: In "Make bibliography" of Jstyles, the generated bibliography section has a heading, whose text and formatting specifications (including paragraph styles) are variable and taken from the jstyle that specifies it. This was, hence, indirectly left to the user who would create the jstyle.
> 


Behaviour should be somewhat consistent between Jstyle and CSLStyles. For styling: just use a standard format from libre office. Heading 1 or sthg.
For the name: maybe the title of the bibliography can be put into a named paragraph, so it can be read before recreating the bibliography. As the default I would just use a localized string from JabRef.

**Comment by koppor** [comment]
This issue has to be fixed: https://github.com/subhramit/jabref/issues/13

File: src/main/java/org/jabref/gui/openoffice/OOBibBase.java
Comment (subhramit) [inline src/main/java/org/jabref/gui/openoffice/OOBibBase.java:948]: remove all commented out code

File: src/main/java/org/jabref/logic/openoffice/oocsltext/CSLCitationOOAdapter.java
Comment (subhramit) [inline src/main/java/org/jabref/logic/openoffice/oocsltext/CSLCitationOOAdapter.java:154]: change method names of this class to more relevant ones (check their flow and utility, maybe simplify so that each method deals with one thing so that we can easily come up with non confusing names)

File: src/main/java/org/jabref/logic/openoffice/oocsltext/CSLUpdateBibliography.java
Comment (subhramit) [inline src/main/java/org/jabref/logic/openoffice/oocsltext/CSLUpdateBibliography.java:63]: check if these methods need to be static
would like an opinion from @calixtus

**Review by subhramit — COMMENTED** [review]
self-review

**Comment by subhramit** [comment]
New icon for make/sync bibliography:

355731065-b87484d2-11fe-4764-a5bc-e4eda12a4718

image

File: src/main/resources/fonts/JabRefMaterialDesign.ttf
Comment (Siedlerchr) [inline src/main/resources/fonts/JabRefMaterialDesign.ttf:1]: Why was the font changed?

Comment (Siedlerchr) [reply src/main/resources/fonts/JabRefMaterialDesign.ttf:1]: Ah I see, this was a non stnadard icon?

Comment (subhramit) [reply src/main/resources/fonts/JabRefMaterialDesign.ttf:1]: yes

**Comment by koppor** [comment]
I fixed sub module issues as follows:

```
$ git merge origin/main
$ git checkout main -- src/main/resources/csl-styles
$ ... git commit ... 
$ git push
```

**Comment by subhramit** [comment]
Should fix https://github.com/subhramit/jabref/issues/21 before merge.

**Comment by subhramit** [comment]
> Should fix [subhramit#21](https://github.com/subhramit/jabref/issues/21) before merge.

Also ref. to update via last commit: Irrelevant (as of now) buttons should be disabled when a CSL style is selected.
image

**Comment by subhramit** [comment]
> Should fix [subhramit#21](https://github.com/subhramit/jabref/issues/21) before merge.

fixed

**Comment by ThiloteE** [comment]
For the record: APA and MLA styles now work with Cite-In-Text.
Has to be seen, if https://github.com/JabRef/jabref/pull/11577/commits/6fdd792967e891c9c76bd807ef198d4a520b62ea has undesired effects on other citationstyles.

File: src/main/java/org/jabref/gui/openoffice/OOBibBase.java
Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/openoffice/OOBibBase.java:122]: @subhramit  Check again does the same as

File: src/main/java/org/jabref/gui/openoffice/OOBibBase.java
Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/openoffice/OOBibBase.java:640]: Check agian @subhramit  and other places

File: src/main/java/org/jabref/logic/citationstyle/CitationStyle.java
Comment (Siedlerchr) [reply src/main/java/org/jabref/logic/citationstyle/CitationStyle.java:256]: ping @subhramit

File: src/main/java/org/jabref/logic/citationstyle/CitationStyle.java
Comment (Siedlerchr) [reply src/main/java/org/jabref/logic/citationstyle/CitationStyle.java:244]: TODO: File an upstream issue for suggesting a method for like isNumericStyle

File: src/main/java/org/jabref/gui/openoffice/OOBibBase.java
Comment (subhramit) [reply src/main/java/org/jabref/gui/openoffice/OOBibBase.java]: resolved. used paragraph style

**Comment by subhramit** [comment]
Pending work on this:
1. Acting on chris' reviews
2. https://github.com/subhramit/jabref/issues/23
3. Removing unnecessary static nature of methods

File: src/main/java/org/jabref/gui/openoffice/OOBibBase.java
Comment (subhramit) [reply src/main/java/org/jabref/gui/openoffice/OOBibBase.java]: resolved

File: src/main/java/org/jabref/logic/openoffice/oocsltext/CSLCitationOOAdapter.java
Comment (subhramit) [reply src/main/java/org/jabref/logic/openoffice/oocsltext/CSLCitationOOAdapter.java:289]: refactoring tasks moved to PR-D

File: src/main/java/org/jabref/logic/citationstyle/CitationStyle.java
Comment (subhramit) [reply src/main/java/org/jabref/logic/citationstyle/CitationStyle.java:244]: Since https://github.com/JabRef/jabref/pull/11604 is merged, we now have an xml builder initialized. How about I use stax parsing here as well?

File: src/main/java/org/jabref/gui/openoffice/OOBibBase.java
Comment (subhramit) [reply src/main/java/org/jabref/gui/openoffice/OOBibBase.java:839]: This existed already (look below at removed line 821).
I can replace the variables with the string, but it will become ugly

Comment (subhramit) [reply src/main/java/org/jabref/gui/openoffice/OOBibBase.java:839]: Seems like this exists in the entire file.

File: src/main/java/org/jabref/gui/openoffice/OOBibBase.java
Comment (subhramit) [reply src/main/java/org/jabref/gui/openoffice/OOBibBase.java]: Done

File: src/main/java/org/jabref/logic/citationstyle/CitationStyle.java
Comment (subhramit) [reply src/main/java/org/jabref/logic/citationstyle/CitationStyle.java]: Done. Also changed this one to StAX

File: src/main/java/org/jabref/logic/citationstyle/CitationStyle.java
Comment (subhramit) [reply src/main/java/org/jabref/logic/citationstyle/CitationStyle.java]: resolved

File: src/main/java/org/jabref/gui/openoffice/OOBibBase.java
Comment (subhramit) [reply src/main/java/org/jabref/gui/openoffice/OOBibBase.java]: resloved

File: src/main/java/org/jabref/logic/openoffice/oocsltext/CSLCitationOOAdapter.java
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/openoffice/oocsltext/CSLCitationOOAdapter.java:348]: This method should be part of AuthorLIst, then you can add a proper test

Comment (subhramit) [reply src/main/java/org/jabref/logic/openoffice/oocsltext/CSLCitationOOAdapter.java:348]: There is already an authorsAlpha in BracketedPattern
Since I am using a derived version for CSL, I was planning to keep this one here.
I have plans for refactoring all the formatters into one CSLFormatUtils in next PR, will put this there if alright?

Comment (Siedlerchr) [reply src/main/java/org/jabref/logic/openoffice/oocsltext/CSLCitationOOAdapter.java:348]: Okay, then add this to the todo list for the next PR(s).

**Comment by subhramit** [comment]
### Updated bibliography icon as per discussion:
image
image

### For comparison:
Previous:
image
Original:
image

**Comment by subhramit** [comment]
@Siedlerchr This PR is now complete as per the tasks, feature wishes and cleanup that was targeted.
Waiting for @koppor to test style tab auto-select and DIN 1505-02 style.
Then, you may approve.

**Review by Siedlerchr — APPROVED** [review]
from my point of view this PR is ready to be integrated now. 
More tests and so on will be done in follow up PRs

**Comment by koppor** [comment]
- [ ] https://github.com/subhramit/jabref/issues/17

**Comment by subhramit** [comment]
> * [ ]  [Add support for alphanumeric styles subhramit/jabref#17](https://github.com/subhramit/jabref/issues/17)

Fixed:
https://github.com/subhramit/jabref/issues/17
https://github.com/subhramit/jabref/issues/23
https://github.com/subhramit/jabref/issues/24

**Comment by subhramit** [comment]
Required for proper functioning of alphanumeric styles: https://github.com/JabRef/jabref/pull/11614

**Review by koppor — DISMISSED** [review]

**Review by Siedlerchr — DISMISSED** [review]

**Review by Siedlerchr — APPROVED** [review]

**Review by Siedlerchr — APPROVED** [review]

**Review by Siedlerchr — APPROVED** [review]

**Comment by koppor** [comment]
Next time, please do look into the small red crosses on GitHub!

image

Will be fixed by @Siedlerchr in the main branch.

**Comment by subhramit** [comment]
> Next time, please do look into the small red crosses on GitHub!
> 
> image
> 
> Fixed by @Siedlerchr #11616 (needed to hurry up, because our tests on main branch were red after the merge)

For some reason:
1. First time the merge failed, only fetcher and checkstyle were red.
2. Second time the merge failed, only fetcher test was red (even though architecture and checkstyle error still existed).
3. At no point previously the unit tests were red.

They should have been. Only fetcher and JDK deployment tests started failing the last time.

**Comment by subhramit** [comment]
Look at this:
image
(from https://github.com/JabRef/jabref/pull/11577/commits/cc36acc48e51033231d24d33e8e3010fd0ad03ec or any commit above).

**Comment by subhramit** [comment]
> Look at this: image (from [cc36acc](https://github.com/JabRef/jabref/commit/cc36acc48e51033231d24d33e8e3010fd0ad03ec) or any commit above).

Fixed delayed panic of tests in https://github.com/JabRef/jabref/pull/11615

**Comment by koppor** [comment]
Now, we all learned about the difference of "pull_request" and "pull_request_target" the long way 😅

**Comment by subhramit** [comment]
Major follow-up PR (apart from PR-D): https://github.com/JabRef/jabref/pull/11712
[Adds number re-distribution for citations].

**Comment by koppor** [comment]
> Decision to leave the bibliography title format

@subhramit I hope that when a user changes the text / formatting manually, this change is NOT overwritten, but kept?

Typically, paper templates provide formatting guidelines and names to headings.

**Comment by subhramit** [comment]
> I hope that when a user changes the text / formatting manually, this change is NOT overwritten, but kept?

If by manually, you mean without JabRef, then unfortunately the entire section is detected and re-generated, and that I don't know if it is possible to detect and avoid.

If you are referring to PR https://github.com/JabRef/jabref/pull/12784:
This gives the user flexibility to set the text (any) and formatting (as per popular formats) if it coincides with the templates you are talking about. If the formatting is not in the list, they can always overwrite it manually at the **end** of their work - which is very doable. Point of the PR was to enable JabRef reduce the user's effort a bit more - earlier it was more disadvantageous as both the text and the format were hardcoded, and the user would have to change them at the end anyway (if other text/formats were needed).

**Comment by koppor** [comment]
> If by manually, you mean without JabRef, then unfortunately the entire section is detected and re-generated, and that I don't know if it is possible to detect and avoid.

Yes.

1. Before regeneration: Search for heading, store attributes (style, text)
2. Regenerate
3. Restore style from step 1.

But the solution of https://github.com/JabRef/jabref/pull/12784 also seems to be OK.

**Comment by subhramit** [comment]
> * Before regeneration: Search for heading, store attributes (style, text)
> * Regenerate
> * Restore style from step 1.

Difficult task development-wise using UNO (and would add more lag in auto-refresh mode), compared to tiny effort by user if they just change at the end... 
Can of worms:
If in future we have type-switching across style families, this will get even more complicated (e.g. CSL -> JStyle: Read what's specified in the `.jstyle` file or prioritize user's manual changes and conditionally disable those properties in the jstyle, and so on).

**Comment by subhramit** [comment]
> If in future we have unified type-switching across style families, this will get even more complicated (e.g. CSL -> JStyle: Respect what's specified in the `.jstyle` file or prioritize user's manual changes and conditionally disable those properties in the jstyle, and so on - style-dependent implementations would be needed for a tiny thing).

I think I just invited an additional ADR for whoever works on this.

**Pull request merged**
